### PR TITLE
Smartcard fixes and cleanup

### PR DESCRIPTION
**Pull request by bmiklautz**
This pull request fixes and improves smart card in winpr SCard API (PCSC wrapper) and the smart card channel. It makes the channel basically work again and might fix most of the reported issues.

I've tested against Windows 2012 R2, Windows 7 and Windows 10 with multiple different applications.
Authentication with smart card is untested (yet) as I don't have a setup for it.

This pull request **breaks** the support for redirecting a single smart card (like /smartcard:XXX). This was removed as it extended the SCard API but only for PCSC - causing different behavior on Windows and other platforms. It will be re-added as part of the smart card channel itself (not yet done).

### CODE REVIEW

File: channels/smartcard/client/smartcard_operations.c [hunk 1]
```diff
@@ -1064,23 +1050,36 @@ static LONG smartcard_StatusA_Call(SMARTCARD_DEVICE* smartcard, SMARTCARD_OPERAT
 	LONG status;
 	Status_Return ret = { 0 };
 	DWORD cchReaderLen = 0;
+	DWORD cbAtrLen = 0;
 	LPSTR mszReaderNames = NULL;
 	IRP* irp = operation->irp;
 	Status_Call* call = operation->call;
+	ZeroMemory(ret.pbAtr, 32);
```
Comment (akallabeth) [inline channels/smartcard/client/smartcard_operations.c]: Already done at initialization time

File: channels/smartcard/client/smartcard_pack.c [hunk 2]
```diff
@@ -850,7 +850,7 @@ void smartcard_trace_list_readers_return(SMARTCARD_DEVICE* smartcard, ListReader
 		CopyMemory(mszA, ret->msz, ret->cBytes);
 	}
 
-	for (index = 0; index < length - 2; index++)
+	for (index = 0; index < length - 1; index++)
```
Comment (akallabeth) [inline channels/smartcard/client/smartcard_pack.c]: index should be ```size_t``` just like length.
There are some remaining casts of ```lenght``` to ```int``` which are no longer required (can't comment as not really in changed area)

File: channels/smartcard/client/smartcard_operations.c [hunk 3]
```diff
@@ -1124,19 +1125,34 @@ static LONG smartcard_StatusW_Call(SMARTCARD_DEVICE* smartcard, SMARTCARD_OPERAT
...
 	if (call->cbAtrLen > 32)
 		call->cbAtrLen = 32;
 
+	if (call->fmszReaderNamesIsNULL)
+		cchReaderLen = 0;
+	else
+		cchReaderLen = SCARD_AUTOALLOCATE;
+
 	ret.cbAtrLen = call->cbAtrLen;
 	ZeroMemory(ret.pbAtr, 32);
```
Comment (akallabeth) [inline channels/smartcard/client/smartcard_operations.c]: init with ```= { 0 }``` could do that

**Comment by dodo040** [comment]
I started a simple test after compiling with address sanitizer :

`cmake -C ci/cmake-preloads/config-linux-all.txt -DWITH_GSTREAMER_1_0=OFF -DWITH_SANITIZE_ADDRESS=ON -DWITH_NO_UNDEFINED=OFF -DWITH_XRANDR=OFF .`

    $ ./xfreerdp /d:domain /u:username /v:remoteserver /p:"password" /smartcard
    [12:11:08:310] [6119:6120] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpdr
    [12:11:08:310] [6119:6120] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpsnd
    [12:11:08:310] [6119:6120] [INFO][com.freerdp.client.common.cmdline] - loading channelEx cliprdr
    [12:11:08:475] [6119:6120] [INFO][com.winpr.sspi.Kerberos] - Authenticated to Kerberos v5 via login/password
    [12:11:08:536] [6119:6120] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_BGRX32
    [12:11:08:537] [6119:6120] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_RGB16
    [12:11:08:559] [6119:6120] [INFO][com.winpr.clipboard] - initialized POSIX local file subsystem
    [12:11:08:574] [6119:6131] [INFO][com.freerdp.channels.rdpdr.client] - Loading device service smartcard (static)
    [12:11:09:732] [6119:6131] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)
    =================================================================
    ==6119==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x603000000000 at pc 0x0000004b46fc bp 0x7ff381359470 sp 0x7ff381358c20
    READ of size 32755 at 0x603000000000 thread T16
        #0 0x4b46fb in __asan_memcpy (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x4b46fb)
        #1 0x7ff396189a10 in Stream_Write /home/dorian/repo-git/bmiklautz/winpr/include/winpr/stream.h:219:2
        #2 0x7ff3961a69ff in smartcard_pack_status_return /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_pack.c:1846:3
        #3 0x7ff396181288 in smartcard_StatusW_Call /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_operations.c:1156:16
        #4 0x7ff396177f35 in smartcard_irp_device_control_call /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_operations.c:1948:13
        #5 0x7ff39616cc49 in smartcard_context_thread /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_main.c:87:19
        #6 0x7ff394fcdc20 in thread_launcher /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:319:7
        #7 0x7ff394be4183 in start_thread /build/eglibc-SvCtMH/eglibc-2.19/nptl/pthread_create.c:312
        #8 0x7ff3944f5ffc in clone /build/eglibc-SvCtMH/eglibc-2.19/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:111

    0x603000000000 is located 16 bytes to the left of 24-byte region         [0x603000000010,0x603000000028)
    allocated by thread T1 here:
        #0 0x4ca5f8 in __interceptor_malloc (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x4ca5f8)
        #1 0x7ff391ea0db2 in CRYPTO_malloc (/lib/x86_64-linux-gnu/libcrypto.so.1.0.0+0x5fdb2)

    Thread T16 created by T8 here:
        #0 0x43396d in pthread_create (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x43396d)
        #1 0x7ff394fcb24a in winpr_StartThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:348:6
        #2 0x7ff394fcac78 in CreateThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:462:8
        #3 0x7ff39616d150 in smartcard_context_new /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_main.c:136:21
        #4 0x7ff396178a0c in smartcard_EstablishContext_Call /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_operations.c:233:14
        #5 0x7ff396177d13 in smartcard_irp_device_control_call /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_operations.c:1820:13
        #6 0x7ff39616e038 in smartcard_process_irp /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_main.c:453:18
        #7 0x7ff3961703eb in smartcard_thread_func /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_main.c:589:18
        #8 0x7ff394fcdc20 in thread_launcher /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:319:7
        #9 0x7ff394be4183 in start_thread /build/eglibc-SvCtMH/eglibc-2.19/nptl/pthread_create.c:312

    Thread T8 created by T6 here:
        #0 0x43396d in pthread_create (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x43396d)
        #1 0x7ff394fcb24a in winpr_StartThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:348:6
        #2 0x7ff394fcca78 in ResumeThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:695:8
        #3 0x7ff39616f221 in smartcard_DeviceServiceEntry /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_main.c:762:2
        #4 0x7ff39607a4f4 in devman_load_device_service /home/dorian/repo-git/bmiklautz/channels/rdpdr/client/devman.c:193:9
        #5 0x7ff39606efa7 in rdpdr_process_connect /home/dorian/repo-git/bmiklautz/channels/rdpdr/client/rdpdr_main.c:1076:16
        #6 0x7ff39606de2e in rdpdr_virtual_channel_client_thread /home/dorian/repo-git/bmiklautz/channels/rdpdr/client/rdpdr_main.c:1633:15
        #7 0x7ff394fcdc20 in thread_launcher /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:319:7
        #8 0x7ff394be4183 in start_thread /build/eglibc-SvCtMH/eglibc-2.19/nptl/pthread_create.c:312

    Thread T6 created by T1 here:
        #0 0x43396d in pthread_create (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x43396d)
        #1 0x7ff394fcb24a in winpr_StartThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:348:6
        #2 0x7ff394fcac78 in CreateThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:462:8
        #3 0x7ff39606d269 in rdpdr_virtual_channel_event_connected /home/dorian/repo-git/bmiklautz/channels/rdpdr/client/rdpdr_main.c:1705:24
        #4 0x7ff39606cd20 in rdpdr_virtual_channel_init_event_ex /home/dorian/repo-git/bmiklautz/channels/rdpdr/client/rdpdr_main.c:1794:17
        #5 0x7ff395a5c7dc in freerdp_channels_post_connect /home/dorian/repo-git/bmiklautz/libfreerdp/core/client.c:371:4
        #6 0x7ff395a53e03 in freerdp_connect /home/dorian/repo-git/bmiklautz/libfreerdp/core/freerdp.c:217:14
        #7 0x5738cc in xf_client_thread /home/dorian/repo-git/bmiklautz/client/X11/xf_client.c:1458:11
        #8 0x7ff394fcdc20 in thread_launcher /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:319:7
        #9 0x7ff394be4183 in start_thread /build/eglibc-SvCtMH/eglibc-2.19/nptl/pthread_create.c:312

    Thread T1 created by T0 here:
        #0 0x43396d in pthread_create (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x43396d)
        #1 0x7ff394fcb24a in winpr_StartThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:348:6
        #2 0x7ff394fcac78 in CreateThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:462:8
        #3 0x5688b2 in xfreerdp_client_start /home/dorian/repo-git/bmiklautz/client/X11/xf_client.c:1678:22
        #4 0x7ff396038d9d in freerdp_client_start /home/dorian/repo-git/bmiklautz/client/common/client.c:128:9
        #5 0x575347 in main /home/dorian/repo-git/bmiklautz/client/X11/cli/xfreerdp.c:71:2
        #6 0x7ff394419f44 in __libc_start_main /build/eglibc-SvCtMH/eglibc-2.19/csu/libc-start.c:287

    SUMMARY: AddressSanitizer: heap-buffer-overflow (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x4b46fb) in __asan_memcpy
    Shadow bytes around the buggy address:
      0x0c067fff7fb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
      0x0c067fff7fc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
      0x0c067fff7fd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
      0x0c067fff7fe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
      0x0c067fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    =>0x0c067fff8000:[fa]fa 00 00 00 fa fa fa 00 00 00 fa fa fa 00 00
      0x0c067fff8010: 00 fa fa fa 00 00 00 fa fa fa 00 00 00 fa fa fa
      0x0c067fff8020: 00 00 00 fa fa fa 00 00 00 fa fa fa 00 00 00 fa
      0x0c067fff8030: fa fa 00 00 00 fa fa fa 00 00 00 fa fa fa 00 00
      0x0c067fff8040: 00 fa fa fa 00 00 00 fa fa fa 00 00 00 fa fa fa
      0x0c067fff8050: 00 00 00 fa fa fa 00 00 00 fa fa fa 00 00 00 fa
    Shadow byte legend (one shadow byte represents 8 application bytes):
      Addressable:           00
      Partially addressable: 01 02 03 04 05 06 07 
      Heap left redzone:       fa
      Heap right redzone:      fb
      Freed heap region:       fd
      Stack left redzone:      f1
      Stack mid redzone:       f2
      Stack right redzone:     f3
      Stack partial redzone:   f4
      Stack after return:      f5
      Stack use after scope:   f8
      Global redzone:          f9
      Global init order:       f6
      Poisoned by user:        f7
      Container overflow:      fc
      Array cookie:            ac
      Intra object redzone:    bb
      ASan internal:           fe
      Left alloca redzone:     ca
      Right alloca redzone:    cb
    ==6119==ABORTING

**Comment by bmiklautz** [comment]
@dodo040 how/what did you test exactly - I compiled with -DWITH_SANITIZE_ADDRESS=ON as well but haven't seen this.

**Comment by bmiklautz** [comment]
@dodo040 and just to make sure you are on the right branch (not master of my repo)?

**Comment by dodo040** [comment]
@bmiklautz I mean I started the program the same way I do usually. 
Note that without address sanitizer it works :

    $ ./xfreerdp /d:domain /u:username /v:remoteserver /p:"password" /smartcard
    [12:40:22:346] [17668:17669] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpdr
    [12:40:22:346] [17668:17669] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpsnd
    [12:40:22:346] [17668:17669] [INFO][com.freerdp.client.common.cmdline] - loading channelEx cliprdr
    [12:40:22:447] [17668:17669] [INFO][com.winpr.sspi.Kerberos] - Authenticated to Kerberos v5 via login/password
    [12:40:22:639] [17668:17669] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_BGRX32
    [12:40:22:639] [17668:17669] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_RGB16
    [12:40:22:659] [17668:17669] [INFO][com.winpr.clipboard] - initialized POSIX local file subsystem
    [12:40:22:662] [17668:17678] [INFO][com.freerdp.channels.rdpdr.client] - Loading device service smartcard (static)
    [12:40:23:246] [17668:17678] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)
    [12:40:24:896] [17668:17678] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)
    [12:40:24:897] [17668:17678] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)
    [12:40:29:863] [17668:17669] [INFO][com.freerdp.client.x11] - Closed from X11
    [12:40:29:864] [17668:17706] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetStatusChangeW (0x000900A4), status: SCARD_E_CANCELLED (0x80100002)
    [12:40:29:865] [17668:17680] [ERROR][com.freerdp.channels.rdpdr.client] - pVirtualChannelWriteEx failed with CHANNEL_RC_BAD_CHANNEL_HANDLE [00000007]
    [12:40:29:865] [17668:17680] [ERROR][com.freerdp.channels.smartcard.client] - smartcard_complete_irp failed with error 7!
    [12:40:29:865] [17668:17712] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetStatusChangeW (0x000900A4), status: SCARD_E_CANCELLED (0x80100002)


Yes I am on the right branch smartcard-cleanup.

**Comment by bmiklautz** [comment]
@dodo040 can't reproduce but would you try with the following patch applied?
```
diff --git a/channels/smartcard/client/smartcard_operations.c b/channels/smartcard/client/smartcard_operations.c
index d0ed44f4d..60c6469a0 100644
--- a/channels/smartcard/client/smartcard_operations.c
+++ b/channels/smartcard/client/smartcard_operations.c
@@ -1135,7 +1135,8 @@ static LONG smartcard_StatusW_Call(SMARTCARD_DEVICE* smartcard, SMARTCARD_OPERAT
        else
                cchReaderLen = SCARD_AUTOALLOCATE;
 
-       ret.cbAtrLen = call->cbAtrLen;
+       cbAtrLen = call->cbAtrLen;
+
        ZeroMemory(ret.pbAtr, 32);
        status = ret.ReturnCode = SCardStatusW(operation->hCard,
                                               call->fmszReaderNamesIsNULL ? NULL : (LPWSTR) &mszReaderNames,
@@ -1147,7 +1148,7 @@ static LONG smartcard_StatusW_Call(SMARTCARD_DEVICE* smartcard, SMARTCARD_OPERAT
                if (!call->fmszReaderNamesIsNULL)
                        ret.mszReaderNames = (BYTE*) mszReaderNames;
 
-               ret.cBytes = cchReaderLen * 2;
+               ret.cBytes = cchReaderLen;
 
                if (call->cbAtrLen)
                        ret.cbAtrLen = cbAtrLen;

```

**Comment by dodo040** [comment]
@bmiklautz after applying the patch with address sanitizer I can connect to my remote session but no smartcard is detected although all the smartcard readers are well redirected.

But if I'm playing to remove and insert the smartcard into the reader it ended by a failure : 

    $ ./xfreerdp /d:domain /u:username /v:remoteserver /p:"password" /smartcard
    [15:22:04:629] [7246:7247] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpdr
    [15:22:04:629] [7246:7247] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpsnd
    [15:22:04:629] [7246:7247] [INFO][com.freerdp.client.common.cmdline] - loading channelEx cliprdr
    [15:22:04:734] [7246:7247] [INFO][com.winpr.sspi.Kerberos] - Authenticated to Kerberos v5 via login/password
    [15:22:04:836] [7246:7247] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_BGRX32
    [15:22:04:836] [7246:7247] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_RGB16
    [15:22:04:865] [7246:7247] [INFO][com.winpr.clipboard] - initialized POSIX local file subsystem
    [15:22:04:868] [7246:7256] [INFO][com.freerdp.channels.rdpdr.client] - Loading device service smartcard (static)
    [15:22:05:535] [7246:7256] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)
    [15:22:07:645] [7246:7266] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetStatusChangeW (0x000900A4), status: SCARD_E_CANCELLED (0x80100002)
    [15:22:07:645] [7246:7270] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetStatusChangeW (0x000900A4), status: SCARD_E_CANCELLED (0x80100002)
    [15:22:07:646] [7246:7256] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)
    [15:22:07:646] [7246:7256] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)
    [15:22:17:107] [7246:7288] [ERROR][com.freerdp.channels.smartcard.client] - SCardBeginTransaction failed with error -2146434968
    [15:22:17:107] [7246:7288] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardBeginTransaction (0x000900BC), status: SCARD_W_RESET_CARD (0x80100068)
    [15:22:17:525] [7246:7288] [ERROR][com.freerdp.channels.smartcard.client] - SCardEndTransaction failed with error -2146435050
    [15:22:17:525] [7246:7288] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardEndTransaction (0x000900C0), status: SCARD_E_NOT_TRANSACTED (0x80100016)
    [15:22:17:559] [7246:7288] [WARN][com.freerdp.channels.smartcard.client] - SCardGetAttrib: SCARD_ATTR_ICC_PRESENCE (0x00090300) cbAttrLen: 1
    [15:22:17:559] [7246:7288] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetAttrib (0x000900D8), status: SCARD_E_SHARING_VIOLATION (0x8010000B)
    [15:22:31:193] [7246:7288] [WARN][com.freerdp.channels.smartcard.client] - SCardGetAttrib: SCARD_ATTR_ICC_PRESENCE (0x00090300) cbAttrLen: 1
    [15:22:31:193] [7246:7288] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetAttrib (0x000900D8), status: SCARD_W_REMOVED_CARD (0x80100069)
    [15:22:34:905] [7246:7288] [ERROR][com.freerdp.channels.smartcard.client] - SCardBeginTransaction failed with error -2146434968
    [15:22:34:905] [7246:7288] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardBeginTransaction (0x000900BC), status: SCARD_W_RESET_CARD (0x80100068)
    [15:22:35:947] [7246:7288] [ERROR][com.freerdp.channels.smartcard.client] - SCardEndTransaction failed with error -2146435050
    [15:22:35:947] [7246:7288] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardEndTransaction (0x000900C0), status: SCARD_E_NOT_TRANSACTED (0x80100016)
    [15:22:35:311] [7246:7288] [WARN][com.freerdp.channels.smartcard.client] - SCardGetAttrib: SCARD_ATTR_ICC_PRESENCE (0x00090300) cbAttrLen: 1
    [15:22:35:311] [7246:7288] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetAttrib (0x000900D8), status: SCARD_E_SHARING_VIOLATION (0x8010000B)
    [15:22:35:317] [7246:7288] [WARN][com.freerdp.channels.smartcard.client] - SCardGetAttrib: SCARD_ATTR_ICC_PRESENCE (0x00090300) cbAttrLen: 1
    [15:22:35:317] [7246:7288] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetAttrib (0x000900D8), status: SCARD_E_SHARING_VIOLATION (0x8010000B)
    =================================================================
    ==7246==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x603000000000 at pc 0x0000004b46fc bp 0x7f3f5ecd7470 sp 0x7f3f5ecd6c20
    READ of size 32575 at 0x603000000000 thread T16
        #0 0x4b46fb in __asan_memcpy (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x4b46fb)
        #1 0x7f3f72ad0a10 in Stream_Write /home/dorian/repo-git/bmiklautz/winpr/include/winpr/stream.h:219:2
        #2 0x7f3f72aed9ff in smartcard_pack_status_return /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_pack.c:1846:3
        #3 0x7f3f72ac8280 in smartcard_StatusW_Call /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_operations.c:1157:16
        #4 0x7f3f72abef35 in smartcard_irp_device_control_call /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_operations.c:1949:13
        #5 0x7f3f72ab3c49 in smartcard_context_thread /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_main.c:87:19
        #6 0x7f3f71914c20 in thread_launcher /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:319:7
        #7 0x7f3f7152b183 in start_thread /build/eglibc-SvCtMH/eglibc-2.19/nptl/pthread_create.c:312
        #8 0x7f3f70e3cffc in clone /build/eglibc-SvCtMH/eglibc-2.19/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:111

    0x603000000000 is located 16 bytes to the left of 24-byte region [0x603000000010,0x603000000028)
    allocated by thread T1 here:
        #0 0x4ca5f8 in __interceptor_malloc (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x4ca5f8)
        #1 0x7f3f6e7e7db2 in CRYPTO_malloc (/lib/x86_64-linux-gnu/libcrypto.so.1.0.0+0x5fdb2)

    Thread T16 created by T8 here:
        #0 0x43396d in pthread_create (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x43396d)
        #1 0x7f3f7191224a in winpr_StartThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:348:6
        #2 0x7f3f71911c78 in CreateThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:462:8
        #3 0x7f3f72ab4150 in smartcard_context_new /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_main.c:136:21
        #4 0x7f3f72abfa0c in smartcard_EstablishContext_Call /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_operations.c:233:14
        #5 0x7f3f72abed13 in smartcard_irp_device_control_call /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_operations.c:1821:13
        #6 0x7f3f72ab5038 in smartcard_process_irp /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_main.c:453:18
        #7 0x7f3f72ab73eb in smartcard_thread_func /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_main.c:589:18
        #8 0x7f3f71914c20 in thread_launcher /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:319:7
        #9 0x7f3f7152b183 in start_thread /build/eglibc-SvCtMH/eglibc-2.19/nptl/pthread_create.c:312

    Thread T8 created by T6 here:
        #0 0x43396d in pthread_create (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x43396d)
        #1 0x7f3f7191224a in winpr_StartThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:348:6
        #2 0x7f3f71913a78 in ResumeThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:695:8
        #3 0x7f3f72ab6221 in smartcard_DeviceServiceEntry /home/dorian/repo-git/bmiklautz/channels/smartcard/client/smartcard_main.c:762:2
        #4 0x7f3f729c14f4 in devman_load_device_service /home/dorian/repo-git/bmiklautz/channels/rdpdr/client/devman.c:193:9
        #5 0x7f3f729b5fa7 in rdpdr_process_connect /home/dorian/repo-git/bmiklautz/channels/rdpdr/client/rdpdr_main.c:1076:16
        #6 0x7f3f729b4e2e in rdpdr_virtual_channel_client_thread /home/dorian/repo-git/bmiklautz/channels/rdpdr/client/rdpdr_main.c:1633:15
        #7 0x7f3f71914c20 in thread_launcher /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:319:7
        #8 0x7f3f7152b183 in start_thread /build/eglibc-SvCtMH/eglibc-2.19/nptl/pthread_create.c:312

    Thread T6 created by T1 here:
        #0 0x43396d in pthread_create (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x43396d)
        #1 0x7f3f7191224a in winpr_StartThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:348:6
        #2 0x7f3f71911c78 in CreateThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:462:8
        #3 0x7f3f729b4269 in rdpdr_virtual_channel_event_connected /home/dorian/repo-git/bmiklautz/channels/rdpdr/client/rdpdr_main.c:1705:24
        #4 0x7f3f729b3d20 in rdpdr_virtual_channel_init_event_ex /home/dorian/repo-git/bmiklautz/channels/rdpdr/client/rdpdr_main.c:1794:17
        #5 0x7f3f723a37dc in freerdp_channels_post_connect /home/dorian/repo-git/bmiklautz/libfreerdp/core/client.c:371:4
        #6 0x7f3f7239ae03 in freerdp_connect /home/dorian/repo-git/bmiklautz/libfreerdp/core/freerdp.c:217:14
        #7 0x5738cc in xf_client_thread /home/dorian/repo-git/bmiklautz/client/X11/xf_client.c:1458:11
        #8 0x7f3f71914c20 in thread_launcher /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:319:7
        #9 0x7f3f7152b183 in start_thread /build/eglibc-SvCtMH/eglibc-2.19/nptl/pthread_create.c:312

    Thread T1 created by T0 here:
        #0 0x43396d in pthread_create (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x43396d)
        #1 0x7f3f7191224a in winpr_StartThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:348:6
        #2 0x7f3f71911c78 in CreateThread /home/dorian/repo-git/bmiklautz/winpr/libwinpr/thread/thread.c:462:8
        #3 0x5688b2 in xfreerdp_client_start /home/dorian/repo-git/bmiklautz/client/X11/xf_client.c:1678:22
        #4 0x7f3f7297fd9d in freerdp_client_start /home/dorian/repo-git/bmiklautz/client/common/client.c:128:9
        #5 0x575347 in main /home/dorian/repo-git/bmiklautz/client/X11/cli/xfreerdp.c:71:2
        #6 0x7f3f70d60f44 in __libc_start_main /build/eglibc-SvCtMH/eglibc-2.19/csu/libc-start.c:287

    SUMMARY: AddressSanitizer: heap-buffer-overflow (/home/dorian/repo-git/bmiklautz/client/X11/xfreerdp+0x4b46fb) in __asan_memcpy
    Shadow bytes around the buggy address:
      0x0c067fff7fb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
      0x0c067fff7fc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
      0x0c067fff7fd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
      0x0c067fff7fe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
      0x0c067fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    =>0x0c067fff8000:[fa]fa 00 00 00 fa fa fa 00 00 00 fa fa fa 00 00
      0x0c067fff8010: 00 fa fa fa 00 00 00 fa fa fa 00 00 00 fa fa fa
      0x0c067fff8020: 00 00 00 fa fa fa 00 00 00 fa fa fa 00 00 00 fa
      0x0c067fff8030: fa fa 00 00 00 fa fa fa 00 00 00 fa fa fa 00 00
      0x0c067fff8040: 00 fa fa fa 00 00 00 fa fa fa 00 00 00 fa fa fa
      0x0c067fff8050: 00 00 00 fa fa fa 00 00 00 fa fa fa 00 00 00 fa
    Shadow byte legend (one shadow byte represents 8 application bytes):
      Addressable:           00
      Partially addressable: 01 02 03 04 05 06 07 
      Heap left redzone:       fa
      Heap right redzone:      fb
      Freed heap region:       fd
      Stack left redzone:      f1
      Stack mid redzone:       f2
      Stack right redzone:     f3
      Stack partial redzone:   f4
      Stack after return:      f5
      Stack use after scope:   f8
      Global redzone:          f9
      Global init order:       f6
      Poisoned by user:        f7
      Container overflow:      fc
      Array cookie:            ac
      Intra object redzone:    bb
      ASan internal:           fe
      Left alloca redzone:     ca
      Right alloca redzone:    cb
    ==7246==ABORTING

**Comment by bmiklautz** [comment]
@dodo040 I pushed a fix just this moment. Sorry to bother you but would you retry?

**Comment by dodo040** [comment]
Way better for me :) LGTM now.

**Comment by bmiklautz** [comment]
@dodo040 nice! Thanks for testing.

**Comment by qxlsz** [comment]
I'm in linux machine 
tried to send APDU
and when i see pcscd daemon, i can see that the smart card is powering and unpowering automatically. what needs to be done to fix this issue ?

**Pull request merged**